arch-design/api/UnifiedAPI.md

```diff
@@ -0,0 +1,208 @@
+
+
+The below [sequence diagrams](#Sequence) can be referred to see the end-to-end flow in more detail.
+
+Please find the details of proposed diagram
+#### Fig. I: The proposed architecture diagram
+![](resources/unified-3.png)
+
+### Sequence Diagrams:
+1. POST http://HOST_IP:PORT/v1beta/{tenantId}/object/xxx
+
+  ![](resources/seq-1.png)
+
+2. POST http://HOST_IP:PORT/v1beta/{tenantId}/object/xxx
+
+  ![](resources/seq-2.png)
+
+3. POST http://HOST_IP:PORT/v1beta/{tenantId}/block/xxx
+
+   POST http://HOST_IP:PORT/v1beta/{tenantId}/file/xxx
+
+  ![](resources/seq-3.png)
+
+4. GET http://HOST_IP:PORT/v1beta/{tenantId}/object/xxx
+
+   GET http://HOST_IP:PORT/v1beta/{tenantId}/{block or file}/xxx
+
+  ![](resources/Get-se-1.png)
+
+
+### Data model impact
+
+New data model will be added to the API server and Controller module. Here are some examples:
+
+```
+type Backend struct {
+	Id         bson.ObjectId `json:"id,omitempty" bson:"_id,omitempty"`
+	TenantId   string        `json:"tenantId,omitempty" bson:"tenantId,omitempty"`
+	UserId     string        `json:"userId,omitempty" bson:"userId,omitempty"`
+	Name       string        `json:"name,omitempty" bson:"name,omitempty"`
+	Type       string        `json:"type,omitempty" bson:"type,omitempty"`
+	Region     string        `json:"region,omitempty" bson:"region,omitempty"`
+	Endpoint   string        `json:"endpoint,omitempty" bson:"endpoint,omitempty"`
+	BucketName string        `json:"bucketName,omitempty" bson:"bucketName,omitempty"`
+	Access     string        `json:"access,omitempty" bson:"access,omitempty"`
+	Security   string        `json:"security,omitempty" bson:"security,omitempty"`
+}
+
+type Bucket struct {
+	Name               string `xml:"Name"`
+	CreateTime         string `xml:"CreateTime"`
+	LocationConstraint string `xml:"LocationConstraint"`
+	VersionOpts        VersioningConfiguration
+	SSEOpts            SSEConfiguration
+}
+```
+
+The full list of all data model can be referred from here:
+1. [Backend service](https://github.com/sodafoundation/multi-cloud/tree/master/backend/pkg/model)
+2. [S3-Service](https://github.com/sodafoundation/multi-cloud/tree/master/s3/pkg/model)
+
+
+### REST API impact
+All these below multi-cloud API's will be added to API server(of hotpot)
+
+##### S3-Service specific:
+
+* PutBucket
+* GetBucket
+* HeadBucket
+* DeleteBucket
+* PutBucketACL
+* GetBucketACL
+* PutBucketCors
+* DeleteBucketCors
+* GetBucketCors
+* PutObject
+* PutObjectCopy
+* GetObject
+* HeadObject
+* DeleteObject
+* Multi-Object Delete
+* PutObjectACL
+* GetObjectACL
+* MultipartUpload
+* GetService
+* Put Bucket Website
+* Get Bucket Website
+* Delete Bucket Website
+* Put Bucket Versioning
+* Get Bucket Versioning
+
+##### Backend Service Specific
+
+* CreateBackend
+* UpdateBackend
+* DeleteBackend
+* GetBackend
+* ListBackend
+
+For list of all multi-cloud API's click [here](https://github.com/sodafoundation/multi-cloud/tree/master/openapi-spec)
+
+### Security impact
```

Comment:
All our S3 APIs are authenticated with V4 auth at the backend. This V4 auth is dependent on the Host URI. Since the API paths will change the V4 auth will have to be tested and this may impact the dashboard and may need some changes to how the hash is generated.

Reply:
Are we changing the backend APIs? they should not change ideally (unless they need to be) 
I think the concern here is different, Like now the APIs like http://xyz.com/backend or HTTP://xyz.com/storage are there so how can we decide if the URLs should be signed or not. If signed, will backend(dock) and backend(multi-cloud) handle it?